FWIO/RNTuple/plugins/RNTupleOutputFile.cc

```diff
@@ -0,0 +1,583 @@
+
+    if (not dropMetaData_) {
+      for (auto const& p : iProducts) {
+        auto h = iEvent.getByToken(p.get_, p.desc_->unwrappedTypeID());
+        if (h.isValid()) {
+          auto prov = h.provenance()->productProvenance();
+          if (not prov) {
+            prov = iEvent.productProvenanceRetrieverPtr()->branchIDToProvenance(p.desc_->originalBranchID());
+          }
+          if (prov) {
+            insertProductProvenance(*prov, provenanceToKeep);
+          }
+        }
+      }
+    }
+    return std::vector<StoredProductProvenance>(provenanceToKeep.begin(), provenanceToKeep.end());
+  }
+
+  bool RNTupleOutputFile::insertProductProvenance(ProductProvenance const& iProv,
+                                                  std::set<StoredProductProvenance>& oToInsert) {
+    StoredProductProvenance toStore;
+    toStore.branchID_ = iProv.branchID().id();
+    auto itFound = oToInsert.find(toStore);
+    if (itFound == oToInsert.end()) {
+      //get the index to the ParentageID or insert a new value if not already present
+      auto i = parentageIDs_.emplace(iProv.parentageID(), static_cast<unsigned int>(parentageIDs_.size()));
+      toStore.parentageIDIndex_ = i.first->second;
+      if (toStore.parentageIDIndex_ >= parentageIDs_.size()) {
+        throw edm::Exception(errors::LogicError)
+            << "RootOutputFile::insertProductProvenance\n"
+            << "The parentage ID index value " << toStore.parentageIDIndex_
+            << " is out of bounds.  The maximum value is currently " << parentageIDs_.size() - 1 << ".\n"
+            << "This should never happen.\n"
+            << "Please report this to the framework developers.";
+      }
+
+      oToInsert.insert(toStore);
+      return true;
+    }
+    return false;
+  }
+
+  void RNTupleOutputFile::insertAncestorsProvenance(ProductProvenance const& iProv,
+                                                    ProductProvenanceRetriever const& iMapper,
+                                                    std::set<StoredProductProvenance>& oToKeep) {
+    std::vector<BranchID> const& parentIDs = iProv.parentage().parents();
+    for (auto const& parentID : parentIDs) {
+      branchesWithStoredHistory_.insert(parentID);
+      ProductProvenance const* info = iMapper.branchIDToProvenance(parentID);
+      if (info) {
+        if (insertProductProvenance(*info, oToKeep)) {
+          //haven't seen this one yet
+          insertAncestorsProvenance(*info, iMapper, oToKeep);
+        }
+      }
+    }
+  }
+
+  void RNTupleOutputFile::write(EventForOutput const& e) {
+    {
+      auto rentry = events_->CreateEntry();
+      rentry->BindRawPtr(*eventAuxField_, const_cast<void*>(static_cast<void const*>(&(e.eventAuxiliary()))));
+      rentry->BindRawPtr(*eventSelField_, const_cast<void*>(static_cast<void const*>(&(e.eventSelectionIDs()))));
+      rentry->BindRawPtr(*branchListField_, const_cast<void*>(static_cast<void const*>(&(e.branchListIndexes()))));
```

Comment:
Can these be
```suggestion
      rentry->BindRawPtr(*eventAuxField_, static_cast<void*>(&(e.eventAuxiliary())));
      rentry->BindRawPtr(*eventSelField_, static_cast<void*>(&(e.eventSelectionIDs())));
      rentry->BindRawPtr(*branchListField_, static_cast<void*>(&(e.branchListIndexes())));
```
?

Reply:
These accessors all return const objects.